### PR TITLE
Add a FeaturesCollection constructor that supports initial capacity

### LINKED ISSUE #31249: New FeatureCollection constructor for initialCapacity
## Background and Motivation
Some background can be found here, dictionary resizes were fairly visible in traces https://github.com/dotnet/aspnetcore/issues/31240

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http.Features
{
    public class FeatureCollection
    {
+       public FeatureCollection(int initialCapacity);
    }
}
```

## Usage Examples
Initial place this will be used is

``` C#
        public DefaultHttpContext()
            : this(new FeatureCollection(10)) // right here
        {
            Features.Set<IHttpRequestFeature>(new HttpRequestFeature());
            Features.Set<IHttpResponseFeature>(new HttpResponseFeature());
            Features.Set<IHttpResponseBodyFeature>(new StreamResponseBodyFeature(Stream.Null));
            // At least some slots left before resize for storing RequestAborted feature, Services and other such things. 
        }
```

The initial capacity of 10 is just an example, say 5 or 7 would also be fine for me.

## Alternative Designs
This could also be handled through an internal constructor.

## Risks
It can be viewed as api bloat if it provides minimal overall value.

### PR DESCRIPTION
**Pull request by mareklinka**
**PR Title**
Add a FeaturesCollection constructor that supports initial capacity

**PR Description**
The `initialCapacity` constructor is added and validates its argument to be `>= 0` . The value is later used when the `_features` dictionary is constructed. The default value remains 0.

The new constructor is used by the default `DefaultHttpContext` ctor with an arbitrary value of 10.

There are currently no tests for this functionality - the actual capacity of the underlying dictionary is not easily observable from the outside and other tests (e.g. validation) seem trivial. If some tests are desired, I can add them in another commit.

Fixes #31249

### CODE REVIEW

File: src/Http/Http/src/DefaultHttpContext.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ public sealed class DefaultHttpContext : HttpContext
         /// Initializes a new instance of the <see cref="DefaultHttpContext"/> class.
         /// </summary>
         public DefaultHttpContext()
-            : this(new FeatureCollection())
+            : this(new FeatureCollection(10))
```
Comment (davidfowl) [inline src/Http/Http/src/DefaultHttpContext.cs:47]: Why 10?

Comment (mareklinka) [reply src/Http/Http/src/DefaultHttpContext.cs:47]: That's taken from the discussion in the original issue, below usage examples. Can we come up with a better number?

Comment (mareklinka) [reply src/Http/Http/src/DefaultHttpContext.cs:47]: Another option would be to add a new constructor to `DefaultHttpContext` as well.

Comment (davidfowl) [reply src/Http/Http/src/DefaultHttpContext.cs:47]: 10 seems reasonable based on common features used based on this list https://github.com/dotnet/aspnetcore/tree/main/src/Http/Http.Features/src. Add a comment, put it in a constant.

Comment (mareklinka) [reply src/Http/Http/src/DefaultHttpContext.cs]: Done.

**Review by davidfowl — APPROVED** [review]

File: src/Http/Http.Features/src/FeatureCollection.cs [hunk 2]
```diff
@@ -73,7 +89,7 @@ public virtual int Revision
 
                 if (_features == null)
                 {
-                    _features = new Dictionary<Type, object>();
+                    _features = new Dictionary<Type, object>(_initialCapacity);
```
Comment (Tratcher) [inline src/Http/Http.Features/src/FeatureCollection.cs:92]: This means that in the default case we're now passing in 0, even though we know we're about to add at least one item. How well does Dictionary handle that?

Comment (NinoFloris) [reply src/Http/Http.Features/src/FeatureCollection.cs:92]: Last I checked dictionary starts at 0 either way and only after the first add allocates space to the next nearest prime from a predefined list, which is 3, then 7 and so on.

Comment (davidfowl) [reply src/Http/Http.Features/src/FeatureCollection.cs:92]: [Dictionary's default capacity is 0](https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L39-L45)

**Pull request merged**